s4e-backend/src/main/java/pl/cyfronet/s4e/license/LicensePermissionEvaluator.java

```diff
@@ -0,0 +1,88 @@
+package pl.cyfronet.s4e.license;
+
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+import pl.cyfronet.s4e.bean.Product;
+import pl.cyfronet.s4e.bean.Scene;
+import pl.cyfronet.s4e.data.repository.ProductRepository;
+import pl.cyfronet.s4e.data.repository.SceneRepository;
+import pl.cyfronet.s4e.license.types.ProductGranularLicense;
+import pl.cyfronet.s4e.license.types.TimestampGranularLicense;
+import pl.cyfronet.s4e.security.AppUserDetails;
+
+import java.util.Map;
+
+@RequiredArgsConstructor
+public class LicensePermissionEvaluator {
+    private final Map<Product.AccessType, ProductGranularLicense> productLicenses;
+    private final ProductRepository productRepository;
+    private final SceneRepository sceneRepository;
+
+    public boolean allowProductRead(Long productId, Object principal) {
+        Product product = productRepository.findById(productId).orElse(null);
+        if (product == null) {
+            return true;
+        }
+        AppUserDetails userDetails = getNullableUserDetails(principal);
+        return decide(userDetails, product);
+    }
+
+    public boolean allowSceneRead(Long sceneId, Object principal) {
+        Scene scene = sceneRepository.findByIdFetchProduct(sceneId).orElse(null);
+        if (scene == null) {
+            return true;
+        }
+        AppUserDetails userDetails = getNullableUserDetails(principal);
+        return decide(userDetails, scene);
+    }
+
+    /**
+     * Decide whether to allow access given permissions.
+     *
+     * @param userDetails the user details
+     * @param product product to authorize
+     * @return If given the permissions from userDetails to allow read access to product
+     */
+    private boolean decide(AppUserDetails userDetails, Product product) {
+        ProductGranularLicense license = productLicenses.get(product.getAccessType());
+        return license.canRead(product, userDetails);
+    }
+
+    /**
+     * Decide whether to allow access given permissions.
+     * <p>
+     * Access is allowed when user can read product and the scene.
+     * <p>
+     * If product's license is not timestamp granular then only product access is verified.
+     *
+     * @param userDetails the user details
+     * @param scene scene to authorize
+     * @return If given the permissions from userDetails to allow read access to scene
+     */
+    private boolean decide(AppUserDetails userDetails, Scene scene) {
+        Product product = scene.getProduct();
+        ProductGranularLicense license = productLicenses.get(product.getAccessType());
+
+        // Verify access to product.
```

Comment:
I don't think so, I already described it in here: https://github.com/cyfronet-fid/sat4envi/pull/778#discussion_r504532433

Reply:
In such a case may be a variable could solve the context. What I mean, you could add instead of the comment variable

```
val hasAccessToProduct = decide(userDetails, product);
if (!hasAccessToProduct) { return false; }

Reply:
I still prefer the narrative in comments.